Docs: Add repository tutorial based on metadata API

lukpueh opened the pull request:
*[EDIT 11/23/2021: remove draft status and update PR description]*
*[EDIT 11/25/2021: remove reviewer question about test, the PR includes one]*

Fixes #1673

**Description**

Add an excessively commented python script to demonstrate repository creation/operation using only the low-level metadata API.

**Context**
As 'repository_tool' and 'repository_lib' are being deprecated, repository metadata must to be created and maintained *manually* using the low-level Metadata API. The example code in these files shall serve as temporary replacement until a new repository tool is available.

**Contents**
 - creation of top-level metadata
 - target file handling
 - consistent snapshots
 - key management
 - top-level delegation and signing thresholds
 - target delegation
 - in-band and out-of-band metadata signing
 - writing and reading metadata files
 - root key rotation

**Questions for reviewers**
 * Is this readable enough? Should this rather be a markdown document?
 * Is it too broad? Should we move some usage scenarios to different files?
   E.g. hashed bin delegation will be in a separate file (see follow-up PR)
 * How can we demonstrate validity? With a TrustedMetadata example?
 * How can we demonstrate usage? With a client example follow-up?

**Please verify and check that the pull request fulfills the following requirements:**
- [X] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [X] Docs have been added for the bug fix or new feature

sechkova commented on examples/repo_example/basic_repo.py:
```diff
@@ -0,0 +1,403 @@
+
+# Common fields
+# -------------
+# All roles have the same metadata container format. On the top-most level,
```
Maybe we can somehow point to the Metadata class here?

sechkova commented on examples/repo_example/basic_repo.py:
```diff
@@ -0,0 +1,403 @@
+# from which a client would download the target file.
+target_local_path = os.path.abspath(__file__)
+target_file_path = os.path.basename(__file__)
+target_file_info = TargetFile.from_file(target_file_path, target_local_path)
```
I have no good suggestion for improvement but these paths seem confusing. I admit I had to read it twice even though I was supposed to know what is going on :grin: Maybe an example path and/vs URL could help.

lukpueh replied:
I can see that how it confusing. I'll try to better explain/exemplify those two paths.

sechkova commented on examples/repo_example/basic_repo.py:
```diff
@@ -0,0 +1,403 @@
+# The timestamp role guarantees freshness of the repository metadata. It does
+# so by listing the latest snapshot (which in turn lists all the latest
+# targets) metadata. Choosing a short expiration interval, requires clients to
+# often re-download timestamp and thus immediately detect if new target files
```
Doesn't the client always try to download timestamp?

lukpueh replied:
Ha! So much for "fully internalized spec". Thanks for catching this. I'll reword.

sechkova commented on examples/repo_example/basic_repo.py:
```diff
@@ -0,0 +1,403 @@
+# Just assume we do out-of-band signing for keys we don't have
+for key in [keys["root"], another_root_key] + new_root_keys:
+    roles["root"].sign(SSlibSigner(key), append=True)
+
```
I think the part "signing with a threshold of old keys AND a threshold of the new keys" needs an emphasis.

lukpueh replied:
➕  I absolutely agree.

sechkova reviewed (commented):
Thanks for the refreshment exercise :) I left some comments on the text itself.
I don't think the example is too broad, having everything at one place seems like a good idea. Maybe in time, more detailed examples can be added for some trickier cases (complex delegations, key rotations ...).

lukpueh commented:
Thanks for your review, @sechkova! I just finalised the example texts and squashed all commits into one. Your comments should be addressed in another commit on top of that for re-review convenience, i.e. 2fc5945.

I'd appreciate another look. Also see the updated PR description, especially the "Questions for reviewers".

sechkova commented on examples/repo_example/basic_repo.py:
```diff
@@ -0,0 +1,397 @@
+
+
+As 'repository_tool' and 'repository_lib' are being deprecated, repository
+metadata must to be created and maintained *manually* using the low-level
```
nit:
```suggestion
metadata must be created and maintained *manually* using the low-level
```

joshuagl replied:
Typo:
```suggestion
metadata must be created and maintained *manually* using the low-level
```

sechkova commented on examples/repo_example/basic_repo.py:
```diff
@@ -103,14 +105,19 @@ def _in(days):
+# configured mirror base URL.
+#
+#      |----base URL---||-------target path-------|
+# e.g. tuf-examples.org/repo_example/basic_repo.py
```
I like it!

lukpueh commented:
I just added very basic testing that runs the example script using `exec` and checks if the metadata files were created as expected. It currently only checks for file existence, but it wouldn't be hard to add more thorough inspection of the created files.

Note that the test class is intended to be re-used for a similar example script that I have in the pipe.

sechkova commented on examples/repo_example/basic_repo.py:
```diff
@@ -0,0 +1,397 @@
+        version=1,
+        spec_version=SPEC_VERSION,
+        expires=_in(31),
+        meta={"targets.json": MetaFile(1)},
```
```suggestion
        meta={"targets.json": MetaFile(version=1)},
```
It is a nit but since it is an example ...

joshuagl replied:
Nit: aids readability IMHO
```suggestion
        meta={"targets.json": MetaFile(version=1)},
```

joshuagl commented on examples/repo_example/basic_repo.py:
```diff
@@ -0,0 +1,397 @@
+As 'repository_tool' and 'repository_lib' are being deprecated, repository
+metadata must to be created and maintained *manually* using the low-level
+Metadata API. The example code in this file shall serve as temporary
+replacement until a new repository tool is available.
```
WDYT to:
```suggestion
Metadata API. The example code in this file demonstrates how to
implement similar functionality to that of the legacy 'repository_tool'
and 'repository_lib' until a new repository implementation is available.
```

Indicates it's an example (and that a user should think before copy/paste verbatim) and doesn't promise to provide a new repository *tool*.

joshuagl commented on examples/repo_example/basic_repo.py:
```diff
@@ -0,0 +1,397 @@
+#
+# The 'version', bumped on each metadata change, is used to reference metadata
+# from within other metadata, and thus allows for repository consistency in
+# addition to protecting against rollback attacks.
```
Is bumped obvious enough a term? If not, perhaps:
```suggestion
# The 'version', which is incremented on each metadata change, is used to
# reference metadata from within other metadata, and thus allows for repository
# consistency in addition to protecting against rollback attacks.
```

joshuagl commented on examples/repo_example/basic_repo.py:
```diff
@@ -0,0 +1,397 @@
+# from within other metadata, and thus allows for repository consistency in
+# addition to protecting against rollback attacks.
+#
+# The expiry date, protects against freeze attacks and allows for implicit key
```
consistency:
```suggestion
# The 'expiry' date protects against freeze attacks and allows for implicit key
```

lukpueh replied:
The field name is 'expires'. Should we write *"The 'expires' date... "*? Does consistency beat grammar? :D 
Regardless, I'll remove the stray comma.

joshuagl replied:
🤦 of course.

I think it is better to bind the comments to the variables, maybe:
"The date the metadata 'expires'..." ?

joshuagl commented on examples/repo_example/basic_repo.py:
```diff
@@ -0,0 +1,397 @@
+# volatility of a role and how easy it is to re-sign them. Highly volatile
+# roles (timestamp, snapshot, delegated targets) usually have shorter
+# expiration intervals, whereas roles that change less and might use offline
+# keys (root, top-level targets) can expire less frequently.
```
Nit: use a similar phrase?
```suggestion
# keys (root, top-level targets) may have longer expiration intervals.
```

joshuagl commented on examples/repo_example/basic_repo.py:
```diff
@@ -0,0 +1,397 @@
+local_path = Path(__file__)
+target_path = f"{local_path.parts[-2]}/{local_path.parts[-1]}"
+
+target_file_info = TargetFile.from_file(target_path, local_path.resolve())
```
Nit: might be cleaner to resolve() earlier, i.e. ln 116. It looks a bit funny doing operations on `local_path` to create `target_path` and only later resolving the full path. I understand why this works, but had to read twice to make sure it did.

lukpueh replied:
Agreed. Will update!

joshuagl commented on examples/repo_example/basic_repo.py:
```diff
@@ -0,0 +1,397 @@
+    Snapshot(
+        version=1,
+        spec_version=SPEC_VERSION,
+        expires=_in(31),
```
Observation: the comments earlier suggest that snapshot is usually a shorter expiration interval than targets, but the inverse is true for this code.

lukpueh replied:
Very good observation! My (admittedly not obvious) thinking here was that the top-level targets in the example is used *like* delegated targets, i.e. to list target files, and thus might change often, and thus might expire more quickly than a top-level target file that only delegates to other targets.

Do you think it is worth including this reasoning in the example, or does it go beyond its scope? I could try to update the comments and maybe also change the expiration interval below, when I make top-level targets a pure delegator.

lukpueh replied:
@joshuagl, do you think 83764c6 is enough?

joshuagl replied:
Seems more than sufficient, thank you!

joshuagl commented on examples/repo_example/basic_repo.py:
```diff
@@ -0,0 +1,397 @@
+# NOTE: We only need the public part to populate root, so it is possible to use
+# out-of-band mechanisms to generate key pairs and only expose the public part
+# to whoever maintains the root role. As a matter of fact, the very purpose of
+# signature thresholds is to avoid that private keys are in one place.
```
```suggestion
# signature thresholds is to avoid having private keys all in one place.
```

sechkova reviewed (approved):
Would be nice if more people provide feedback but you have my approval :grin: 
I also filled in the questionnaire.

>     Is this readable enough? Should this rather be a markdown document?
No strong opinion here, I'm ok with this one
>     Is it too broad? Should we move some usage scenarios to different files?
>     E.g. hashed bin delegation will be in a separate file (see follow-up PR)
Nope, I like it,
>     How can we demonstrate validity? With a TrustedMetadata example?
Seems like the options are either a full Updater with a server or using  TrustedMetadata by providing it with the metadata files as bytes. The second option sounds more interesting to me :) 
>     How can we demonstrate usage? With a client example follow-up?
I have nothing better to propose.

joshuagl commented on examples/repo_example/basic_repo.py:
```diff
@@ -0,0 +1,397 @@
+# Similar to how the root role delegates responsibilities about integrity,
+# consistency and freshness to the corresponding top-level roles, a targets
+# role may further delegate its responsibility for target files (or a subset
+# thereof) to other targets roles. This allows to build a granular trust
+# hierarchy, and further reduces the impact of a single role compromise.
```
```suggestion
# thereof) to other targets roles. This allows creation of a granular trust
# hierarchy, and further reduces the impact of a single role compromise.
```

joshuagl commented on examples/repo_example/basic_repo.py:
```diff
@@ -0,0 +1,397 @@
+# of a unique delegatee name and constraints about the target files the
+# delegatee is responsible for, e.g. a list of path patterns. For details about
+# all configuration parameters see
+# https://theupdateframework.github.io/specification/latest/#delegations for
```
typo:
```suggestion
# https://theupdateframework.github.io/specification/latest/#delegations
```

joshuagl commented on examples/repo_example/basic_repo.py:
```diff
@@ -0,0 +1,397 @@
+        version=1,
+        spec_version=SPEC_VERSION,
+        expires=_in(1),
+        snapshot_meta=MetaFile(1),
```
Nit: same here
```suggestion
        snapshot_meta=MetaFile(version=1),
```

joshuagl commented on tests/test_examples.py:
```diff
@@ -0,0 +1,84 @@
+    """Unit test class for repo example scripts.
+
+    Provides a '_run_example_script' method to run (exec) a script located in
+    the 'repo example' directory.
```
```suggestion
    the 'repo_example' directory.
```

joshuagl commented on tests/test_examples.py:
```diff
@@ -0,0 +1,84 @@
+
+    @classmethod
+    def setUpClass(cls):
+        """Locate and cache repo example dir."""
```
```suggestion
        """Locate and cache repo_example dir."""
```

joshuagl commented on tests/test_examples.py:
```diff
@@ -0,0 +1,84 @@
+    def setUpClass(cls):
+        """Locate and cache repo example dir."""
+        base = Path(__file__).resolve().parents[1]
+        cls.repo_examples_dir = base / "examples" / "repo_example"
```
Should we use `os.path.join()` or do we not want to run this on Windows?

lukpueh replied:
This actually works on Windows too, because `pathlib` overloads the `/` operator to make it act like `join`.

https://docs.python.org/3/library/pathlib.html#operators

joshuagl replied:
Oh, wow. TIL. I cannot decide whether I like this magic or not 😆

joshuagl commented on tests/test_examples.py:
```diff
@@ -0,0 +1,84 @@
+        shutil.rmtree(self.base_test_dir)
+
+    def _run_script_and_assert_files(self, script_name, filenames_created):
+        """Run script in 'repo example' dir and assert that it created the
```
```suggestion
        """Run script in 'repo_example' dir and assert that it created the
```

joshuagl reviewed (approved):
Thanks for this @lukpueh, beautiful work – very clear example and a testament to the Metadata API. 🎉 

> Is this readable enough? Should this rather be a markdown document?

To me this is very readable, appreciate the detail and care here.

> Is it too broad? Should we move some usage scenarios to different files?

LGTM as is. Covers the common cases, I think.

> How can we demonstrate validity? With a TrustedMetadata example?

Implicit in the below? And future work once the proposed repository API validation piece is implemented? i.e. I think this is a solid submission as-is.

> How can we demonstrate usage? With a client example follow-up?

Could be a simple follow-on that runs a http server over the data generated here and uses ngclient to fetch the python file?

jku commented on examples/repo_example/basic_repo.py:
```diff
@@ -0,0 +1,397 @@
+    ),
+)
+
+# Remove target file info from top-level targets (delegatee is now repsonible)
```
```suggestion
# Remove target file info from top-level targets (delegatee is now responsible)
```

jku commented on tests/test_examples.py:
```diff
@@ -0,0 +1,84 @@
+        os.chdir(self.original_cwd)
+        shutil.rmtree(self.base_test_dir)
+
+    def _run_script_and_assert_files(self, script_name, filenames_created):
```
martin is trying to enable mypy (along with the linters) in new test files so argument type annotations would be nice -- but handling these in a follow-up issue is fine IMO.

lukpueh replied:
If you really don't mind, I'll leave this to a follow-up.

jku reviewed (approved):
LGTM, I think this is fine as the stop gap: it's demonstrating the basic elements that a repository needs to handle, without pretending to actually be a repository implementation.

My reaction to the "is it too broad" question is that we shouldn't put a lot of effort into expanding or splitting it (although the hashed bin example should be useful), if we can push the repository library and its example code (or a repository tool) forward instead...

The repository validation question is more interesting... but even there maybe the answer isn't to expand this tutorial code . 

On the usage question: this could be a follow-up issue that is solved when initial versions of both this and the client example are present. Possible solution:
* document how to serve the generated directory over http
* add a "bootstrap from this root.json" option to the client so it can be used with the generated repository

lukpueh commented:
Thanks for the thorough reviews, @sechkova, @joshuagl and @jku! I addressed your comments in  4327214 and  83764c6 (the latter in response to https://github.com/theupdateframework/python-tuf/pull/1685#discussion_r756967300).

May I ask for another quick round of LGTMs? (NOTE: the commits are labeled WIP, because I intend to squash them before merging.)

jku reviewed (approved):
:+1:

joshuagl reviewed (approved):
LGTM, thanks!

MVrachev reviewed (commented):
I just read it and liked the comments and comparing that to `tuf/repository_tool.py` is such an improvement and simplification.
Great work!

Commit:
doc: add repository example based on metadata API

As 'repository_tool' and 'repository_lib' are being deprecated,
repository metadata must to be created and maintained manually
using the low-level Metadata API. The added example code shall
serve as temporary replacement until a new repository tool is
available.

The sample code contains the following repo workflows:
 - creation of top-level metadata
 - target file handling
 - consistent snapshots
 - key management
 - top-level delegation and signing thresholds
 - target delegation
 - in-band and out-of-band metadata signing
 - writing and reading metadata files
 - root key rotation

Co-authored-by: Teodora Sechkova <tsechkova@vmware.com>
Co-authored-by: Joshua Lock <jlock@vmware.com>
Co-authored-by: Jussi Kukkonen <jku@goto.fi>

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
test: add test module for basic repo example

Adds new test module that executes the basic repo example
Python script and checks that it created certain (metadata)
files.

The test module is tailored for testing similar example scripts.

Co-authored-by: Joshua Lock <jlock@vmware.com>
Co-authored-by: Jussi Kukkonen <jku@goto.fi>

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

lukpueh commented:
Thanks, everyone!
